File: apps/mobile/modules/import-sources/ios/Sha256Sink.swift
import CryptoKit
import Foundation

/// Streaming SHA-256, updated per copy chunk so hashing rides the copy's one
/// read. Emits bare lowercase hex; the TS package index is the only place
/// that prefixes `sha256:`.
public struct Sha256Sink {
  private var hasher = SHA256()

  public init() {}

  public mutating func update(_ data: Data) {
    hasher.update(data: data)
  }

  public func finalizeHex() -> String {
    hasher.finalize().map { String(format: "%02x", $0) }.joined()
  }
}

File: apps/mobile/modules/import-sources/android/src/main/java/expo/modules/importsources/Sha256Sink.kt
package expo.modules.importsources

import java.security.MessageDigest

/**
 * Streaming SHA-256 updated per copy chunk. Emits bare lowercase hex; only
 * the TS package index prefixes `sha256:`.
 */
class Sha256Sink {
  private val digest = MessageDigest.getInstance("SHA-256")

  fun update(bytes: ByteArray, length: Int = bytes.size) {
    digest.update(bytes, 0, length)
  }

  fun finalizeHex(): String = digest.digest().joinToString("") { "%02x".format(it) }
}
